src/validation.cpp

```diff
@@ -3622,17 +3622,41 @@ bool BlockManager::LoadBlockIndex(
         vSortedByHeight.push_back(std::make_pair(pindex->nHeight, pindex));
     }
     sort(vSortedByHeight.begin(), vSortedByHeight.end());
+
+    // Find start of assumed-valid region.
+    int first_assumed_valid_height = std::numeric_limits<int>::max();
+
+    for (const auto& [height, block] : vSortedByHeight) {
+        if (block->IsAssumedValid()) {
+            auto chainstates = chainman.GetAll();
+
+            // assert that the assumed valid blocks come from some assumed-valid chainstate.
+            assert(any_of(chainstates.cbegin(), chainstates.cend(),
+                [](auto chainstate) { return chainstate->reliesOnAssumedValid(); }));
+            // assert that the assumed valid blocks will be validated by some not assumed-valid
+            // chainstate.
+            assert(any_of(chainstates.cbegin(), chainstates.cend(),
+                [](auto chainstate) { return !chainstate->reliesOnAssumedValid(); }));
```

Comment:
> This condition should be independent of the loop and could be moved outside?

I don't see why you say that because (i) this condition should only be asserted if we encounter an assumed-valid block, which we need to loop through the index to find out, and (ii) the condition is only asserted once since we break immediately afterwards.

Reply:
I've refactored this code and updated the commentary; hopefully you find it more intuitive now.

Reply:
With "condition" I mean "two chainstates exist".

Thus a loop of the form:

```py
for b in blocks:
  if b.assumed:
    assert len(chains())==2
```

Can be rewritten into:

```py
have_two_chains = len(chains())==2
for b in blocks:
  if b.assumed:
    assert have_two_chains
```

Obviously the performance should be the same here, but evaluating the condition outside the loop might be clearer. If would even allow to add additional checks. (Like asserting that *no* block is assumed if the number of chains is one?)